contrib/perl-json/blib/man3/JSON.3
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "JSON 3"
.TH JSON 3 "2010-09-28" "perl v5.8.8" "User Contributed Perl Documentation"
.SH "NAME"
JSON \- JSON (JavaScript Object Notation) encoder/decoder
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\& use JSON; # imports encode_json, decode_json, to_json and from_json.
.Ve
.PP
.Vb 1
\& # simple and fast interfaces (expect/generate UTF-8)
.Ve
.PP
.Vb 2
\& $utf8_encoded_json_text = encode_json $perl_hash_or_arrayref;
\& $perl_hash_or_arrayref  = decode_json $utf8_encoded_json_text;
.Ve
.PP
.Vb 1
\& # OO-interface
.Ve
.PP
.Vb 1
\& $json = JSON->new->allow_nonref;
.Ve
.PP
.Vb 2
\& $json_text   = $json->encode( $perl_scalar );
\& $perl_scalar = $json->decode( $json_text );
.Ve
.PP
.Vb 1
\& $pretty_printed = $json->pretty->encode( $perl_scalar ); # pretty-printing
.Ve
.PP
.Vb 2
\& # If you want to use PP only support features, call with '-support_by_pp'
\& # When XS unsupported feature is enable, using PP (de|en)code instead of XS ones.
.Ve
.PP
.Vb 1
\& use JSON -support_by_pp;
.Ve
.PP
.Vb 1
\& # option-acceptable interfaces (expect/generate UNICODE by default)
.Ve
.PP
.Vb 2
\& $json_text   = to_json( $perl_scalar, { ascii => 1, pretty => 1 } );
\& $perl_scalar = from_json( $json_text, { utf8  => 1 } );
.Ve
.PP
.Vb 3
\& # Between (en|de)code_json and (to|from)_json, if you want to write
\& # a code which communicates to an outer world (encoded in UTF-8),
\& # recommend to use (en|de)code_json.
.Ve
.SH "VERSION"
.IX Header "VERSION"
.Vb 1
\&    2.26
.Ve
.PP
This version is compatible with \s-1JSON::XS\s0 \fB2.27\fR and later.
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
.Vb 6
\& ************************** CAUTION ********************************
\& * This is 'JSON module version 2' and there are many differences  *
\& * to version 1.xx                                                 *
\& * Please check your applications useing old version.              *
\& *   See to 'INCOMPATIBLE CHANGES TO OLD VERSION'                  *
\& *******************************************************************
.Ve
.PP
\&\s-1JSON\s0 (JavaScript Object Notation) is a simple data format.
See to <http://www.json.org/> and \f(CW\*(C`RFC4627\*(C'\fR(<http://www.ietf.org/rfc/rfc4627.txt>).
.PP
This module converts Perl data structures to \s-1JSON\s0 and vice versa using either
\&\s-1JSON::XS\s0 or \s-1JSON::PP\s0.
.PP
\&\s-1JSON::XS\s0 is the fastest and most proper \s-1JSON\s0 module on \s-1CPAN\s0 which must be
compiled and installed in your environment.
\&\s-1JSON::PP\s0 is a pure-Perl module which is bundled in this distribution and
has a strong compatibility to \s-1JSON::XS\s0.
.PP
This module try to use \s-1JSON::XS\s0 by default and fail to it, use \s-1JSON::PP\s0 instead.
So its features completely depend on \s-1JSON::XS\s0 or \s-1JSON::PP\s0.
.PP
See to \*(L"\s-1BACKEND\s0 \s-1MODULE\s0 \s-1DECISION\s0\*(R".
.PP
To distinguish the module name '\s-1JSON\s0' and the format type \s-1JSON\s0,
the former is quoted by C<> (its results vary with your using media),
and the latter is left just as it is.
.PP
Module name : \f(CW\*(C`JSON\*(C'\fR
.PP
Format type : \s-1JSON\s0
.Sh "\s-1FEATURES\s0"
.IX Subsection "FEATURES"
.IP "* correct unicode handling" 4
.IX Item "correct unicode handling"
This module (i.e. backend modules) knows how to handle Unicode, documents
how and when it does so, and even documents what \*(L"correct\*(R" means.
.Sp
Even though there are limitations, this feature is available since Perl version 5.6.
.Sp
\&\s-1JSON::XS\s0 requires Perl 5.8.2 (but works correctly in 5.8.8 or later), so in older versions
\&\f(CW\*(C`JSON\*(C'\fR sholud call \s-1JSON::PP\s0 as the backend which can be used since Perl 5.005.
.Sp
With Perl 5.8.x \s-1JSON::PP\s0 works, but from 5.8.0 to 5.8.2, because of a Perl side problem,
\&\s-1JSON::PP\s0 works slower in the versions. And in 5.005, the Unicode handling is not available.
See to \*(L"\s-1UNICODE\s0 \s-1HANDLING\s0 \s-1ON\s0 \s-1PERLS\s0\*(R" in \s-1JSON::PP\s0 for more information.
.Sp
See also to \*(L"A \s-1FEW\s0 \s-1NOTES\s0 \s-1ON\s0 \s-1UNICODE\s0 \s-1AND\s0 \s-1PERL\s0\*(R" in \s-1JSON::XS\s0
and \*(L"\s-1ENCODING/CODESET_FLAG_NOTES\s0\*(R" in \s-1JSON::XS\s0.
.IP "* round-trip integrity" 4
.IX Item "round-trip integrity"
When you serialise a perl data structure using only data types supported
by \s-1JSON\s0 and Perl, the deserialised data structure is identical on the Perl
level. (e.g. the string \*(L"2.0\*(R" doesn't suddenly become \*(L"2\*(R" just because
it looks like a number). There \fIare\fR minor exceptions to this, read the
\&\*(L"\s-1MAPPING\s0\*(R" section below to learn about those.
.IP "* strict checking of \s-1JSON\s0 correctness" 4
.IX Item "strict checking of JSON correctness"
There is no guessing, no generating of illegal \s-1JSON\s0 texts by default,
and only \s-1JSON\s0 is accepted as input by default (the latter is a security
feature).
.Sp
See to \*(L"\s-1FEATURES\s0\*(R" in \s-1JSON::XS\s0 and \*(L"\s-1FEATURES\s0\*(R" in \s-1JSON::PP\s0.
.IP "* fast" 4
.IX Item "fast"
This module returns a \s-1JSON::XS\s0 object itself if available.
Compared to other \s-1JSON\s0 modules and other serialisers such as Storable,
\&\s-1JSON::XS\s0 usually compares favourably in terms of speed, too.
.Sp
If not available, \f(CW\*(C`JSON\*(C'\fR returns a \s-1JSON::PP\s0 object instead of \s-1JSON::XS\s0 and
it is very slow as pure\-Perl.
.IP "* simple to use" 4
.IX Item "simple to use"
This module has both a simple functional interface as well as an
object oriented interface interface.
.IP "* reasonably versatile output formats" 4
.IX Item "reasonably versatile output formats"
You can choose between the most compact guaranteed-single-line format possible
(nice for simple line-based protocols), a pure-ASCII format (for when your transport
is not 8\-bit clean, still supports the whole Unicode range), or a pretty-printed
format (for when you want to read that stuff). Or you can combine those features
in whatever way you like.
.SH "FUNCTIONAL INTERFACE"
.IX Header "FUNCTIONAL INTERFACE"
Some documents are copied and modified from \*(L"\s-1FUNCTIONAL\s0 \s-1INTERFACE\s0\*(R" in \s-1JSON::XS\s0.
\&\f(CW\*(C`to_json\*(C'\fR and \f(CW\*(C`from_json\*(C'\fR are additional functions.
.Sh "encode_json"
.IX Subsection "encode_json"
.Vb 1
\&    $json_text = encode_json $perl_scalar
.Ve
.PP
Converts the given Perl data structure to a \s-1UTF\-8\s0 encoded, binary string.
.PP
This function call is functionally identical to:
.PP
.Vb 1
\&    $json_text = JSON->new->utf8->encode($perl_scalar)
.Ve
.Sh "decode_json"
.IX Subsection "decode_json"
.Vb 1
\&    $perl_scalar = decode_json $json_text
.Ve
.PP
The opposite of \f(CW\*(C`encode_json\*(C'\fR: expects an \s-1UTF\-8\s0 (binary) string and tries
to parse that as an \s-1UTF\-8\s0 encoded \s-1JSON\s0 text, returning the resulting
reference.
.PP
This function call is functionally identical to:
.PP
.Vb 1
\&    $perl_scalar = JSON->new->utf8->decode($json_text)
.Ve
.Sh "to_json"
.IX Subsection "to_json"
.Vb 1
\&   $json_text = to_json($perl_scalar)
.Ve
.PP
Converts the given Perl data structure to a json string.
.PP
This function call is functionally identical to:
.PP
.Vb 1
\&   $json_text = JSON->new->encode($perl_scalar)
.Ve
.PP
Takes a hash reference as the second.
.PP
.Vb 1
\&   $json_text = to_json($perl_scalar, $flag_hashref)
.Ve
.PP
So,
.PP
.Vb 1
\&   $json_text = encode_json($perl_scalar, {utf8 => 1, pretty => 1})
.Ve
.PP
equivalent to:
.PP
.Vb 1
\&   $json_text = JSON->new->utf8(1)->pretty(1)->encode($perl_scalar)
.Ve
.PP
If you want to write a modern perl code which communicates to outer world,
you should use \f(CW\*(C`encode_json\*(C'\fR (supposed that \s-1JSON\s0 data are encoded in \s-1UTF\-8\s0).
.Sh "from_json"
.IX Subsection "from_json"
.Vb 1
\&   $perl_scalar = from_json($json_text)
.Ve
.PP
The opposite of \f(CW\*(C`to_json\*(C'\fR: expects a json string and tries
to parse it, returning the resulting reference.
.PP
This function call is functionally identical to:
.PP
.Vb 1
\&    $perl_scalar = JSON->decode($json_text)
.Ve
.PP
Takes a hash reference as the second.
.PP
.Vb 1
\&    $perl_scalar = from_json($json_text, $flag_hashref)
.Ve
.PP
So,
.PP
.Vb 1
\&    $perl_scalar = from_json($json_text, {utf8 => 1})
.Ve
.PP
equivalent to:
.PP
.Vb 1
\&    $perl_scalar = JSON->new->utf8(1)->decode($json_text)
.Ve
.PP
If you want to write a modern perl code which communicates to outer world,
you should use \f(CW\*(C`decode_json\*(C'\fR (supposed that \s-1JSON\s0 data are encoded in \s-1UTF\-8\s0).
.Sh "JSON::is_bool"
.IX Subsection "JSON::is_bool"
.Vb 1
\&    $is_boolean = JSON::is_bool($scalar)
.Ve
.PP
Returns true if the passed scalar represents either JSON::true or
JSON::false, two constants that act like \f(CW1\fR and \f(CW0\fR respectively
and are also used to represent \s-1JSON\s0 \f(CW\*(C`true\*(C'\fR and \f(CW\*(C`false\*(C'\fR in Perl strings.
.Sh "JSON::true"
.IX Subsection "JSON::true"
Returns \s-1JSON\s0 true value which is blessed object.
It \f(CW\*(C`isa\*(C'\fR JSON::Boolean object.
.Sh "JSON::false"
.IX Subsection "JSON::false"
Returns \s-1JSON\s0 false value which is blessed object.
It \f(CW\*(C`isa\*(C'\fR JSON::Boolean object.
.Sh "JSON::null"
.IX Subsection "JSON::null"
Returns \f(CW\*(C`undef\*(C'\fR.
.PP
See \s-1MAPPING\s0, below, for more information on how \s-1JSON\s0 values are mapped to
Perl.
.SH "HOW DO I DECODE A DATA FROM OUTER AND ENCODE TO OUTER"
.IX Header "HOW DO I DECODE A DATA FROM OUTER AND ENCODE TO OUTER"
This section supposes that your perl vresion is 5.8 or later.
.PP
If you know a \s-1JSON\s0 text from an outer world \- a network, a file content, and so on,
is encoded in \s-1UTF\-8\s0, you should use \f(CW\*(C`decode_json\*(C'\fR or \f(CW\*(C`JSON\*(C'\fR module object
with \f(CW\*(C`utf8\*(C'\fR enable. And the decoded result will contain \s-1UNICODE\s0 characters.
.PP
.Vb 4
\&  # from network
\&  my $json        = JSON->new->utf8;
\&  my $json_text   = CGI->new->param( 'json_data' );
\&  my $perl_scalar = $json->decode( $json_text );
.Ve
.PP
.Vb 5
\&  # from file content
\&  local $/;
\&  open( my $fh, '<', 'json.data' );
\&  $json_text   = <$fh>;
\&  $perl_scalar = decode_json( $json_text );
.Ve
.PP
If an outer data is not encoded in \s-1UTF\-8\s0, firstly you should \f(CW\*(C`decode\*(C'\fR it.
.PP
.Vb 5
\&  use Encode;
\&  local $/;
\&  open( my $fh, '<', 'json.data' );
\&  my $encoding = 'cp932';
\&  my $unicode_json_text = decode( $encoding, <$fh> ); # UNICODE
.Ve
.PP
.Vb 4
\&  # or you can write the below code.
\&  #
\&  # open( my $fh, "<:encoding($encoding)", 'json.data' );
\&  # $unicode_json_text = <$fh>;
.Ve
.PP
In this case, \f(CW$unicode_json_text\fR is of course \s-1UNICODE\s0 string.
So you \fBcannot\fR use \f(CW\*(C`decode_json\*(C'\fR nor \f(CW\*(C`JSON\*(C'\fR module object with \f(CW\*(C`utf8\*(C'\fR enable.
Instead of them, you use \f(CW\*(C`JSON\*(C'\fR module object with \f(CW\*(C`utf8\*(C'\fR disable or \f(CW\*(C`from_json\*(C'\fR.
.PP
.Vb 3
\&  $perl_scalar = $json->utf8(0)->decode( $unicode_json_text );
\&  # or
\&  $perl_scalar = from_json( $unicode_json_text );
.Ve
.PP
Or \f(CW\*(C`encode 'utf8'\*(C'\fR and \f(CW\*(C`decode_json\*(C'\fR:
.PP
.Vb 2
\&  $perl_scalar = decode_json( encode( 'utf8', $unicode_json_text ) );
\&  # this way is not efficient.
.Ve
.PP
And now, you want to convert your \f(CW$perl_scalar\fR into \s-1JSON\s0 data and
send it to an outer world \- a network or a file content, and so on.
.PP
Your data usually contains \s-1UNICODE\s0 strings and you want the converted data to be encoded
in \s-1UTF\-8\s0, you should use \f(CW\*(C`encode_json\*(C'\fR or \f(CW\*(C`JSON\*(C'\fR module object with \f(CW\*(C`utf8\*(C'\fR enable.
.PP
.Vb 3
\&  print encode_json( $perl_scalar ); # to a network? file? or display?
\&  # or
\&  print $json->utf8->encode( $perl_scalar );
.Ve
.PP
If \f(CW$perl_scalar\fR does not contain \s-1UNICODE\s0 but \f(CW$encoding\fR\-encoded strings
for some reason, then its characters are regarded as \fBlatin1\fR for perl
(because it does not concern with your \f(CW$encoding\fR).
You \fBcannot\fR use \f(CW\*(C`encode_json\*(C'\fR nor \f(CW\*(C`JSON\*(C'\fR module object with \f(CW\*(C`utf8\*(C'\fR enable.
Instead of them, you use \f(CW\*(C`JSON\*(C'\fR module object with \f(CW\*(C`utf8\*(C'\fR disable or \f(CW\*(C`to_json\*(C'\fR.
Note that the resulted text is a \s-1UNICODE\s0 string but no problem to print it.
.PP
.Vb 6
\&  # $perl_scalar contains $encoding encoded string values
\&  $unicode_json_text = $json->utf8(0)->encode( $perl_scalar );
\&  # or 
\&  $unicode_json_text = to_json( $perl_scalar );
\&  # $unicode_json_text consists of characters less than 0x100
\&  print $unicode_json_text;
.Ve
.PP
Or \f(CW\*(C`decode $encoding\*(C'\fR all string values and \f(CW\*(C`encode_json\*(C'\fR:
.PP
.Vb 3
\&  $perl_scalar->{ foo } = decode( $encoding, $perl_scalar->{ foo } );
\&  # ... do it to each string values, then encode_json
\&  $json_text = encode_json( $perl_scalar );
.Ve
.PP
This method is a proper way but probably not efficient.
.PP
See to Encode, perluniintro.
.SH "COMMON OBJECT-ORIENTED INTERFACE"
.IX Header "COMMON OBJECT-ORIENTED INTERFACE"
.Sh "new"
.IX Subsection "new"
.Vb 1
\&    $json = new JSON
.Ve
.PP
Returns a new \f(CW\*(C`JSON\*(C'\fR object inherited from either \s-1JSON::XS\s0 or \s-1JSON::PP\s0
that can be used to de/encode \s-1JSON\s0 strings.
.PP
All boolean flags described below are by default \fIdisabled\fR.
.PP
The mutators for flags all return the \s-1JSON\s0 object again and thus calls can
be chained:
.PP
.Vb 2
\&   my $json = JSON->new->utf8->space_after->encode({a => [1,2]})
\&   => {"a": [1, 2]}
.Ve
.Sh "ascii"
.IX Subsection "ascii"
.Vb 1
\&    $json = $json->ascii([$enable])
.Ve
.PP
.Vb 1
\&    $enabled = $json->get_ascii
.Ve
.PP
If \f(CW$enable\fR is true (or missing), then the encode method will not generate characters outside
the code range 0..127. Any Unicode characters outside that range will be escaped using either
a single \euXXXX or a double \euHHHH\euLLLLL escape sequence, as per \s-1RFC4627\s0.
.PP
If \f(CW$enable\fR is false, then the encode method will not escape Unicode characters unless
required by the \s-1JSON\s0 syntax or other flags. This results in a faster and more compact format.
.PP
This feature depends on the used Perl version and environment.
.PP
See to \*(L"\s-1UNICODE\s0 \s-1HANDLING\s0 \s-1ON\s0 \s-1PERLS\s0\*(R" in \s-1JSON::PP\s0 if the backend is \s-1PP\s0.
.PP
.Vb 2
\&  JSON->new->ascii(1)->encode([chr 0x10401])
\&  => ["\eud801\eudc01"]
.Ve
.Sh "latin1"
.IX Subsection "latin1"
.Vb 1
\&    $json = $json->latin1([$enable])
.Ve
.PP
.Vb 1
\&    $enabled = $json->get_latin1
.Ve
.PP
If \f(CW$enable\fR is true (or missing), then the encode method will encode the resulting \s-1JSON\s0
text as latin1 (or iso\-8859\-1), escaping any characters outside the code range 0..255.
.PP
If \f(CW$enable\fR is false, then the encode method will not escape Unicode characters
unless required by the \s-1JSON\s0 syntax or other flags.
.PP
.Vb 2
\&  JSON->new->latin1->encode (["\ex{89}\ex{abc}"]
\&  => ["\ex{89}\e\eu0abc"]    # (perl syntax, U+abc escaped, U+89 not)
.Ve
.Sh "utf8"
.IX Subsection "utf8"
.Vb 1
\&    $json = $json->utf8([$enable])
.Ve
.PP
.Vb 1
\&    $enabled = $json->get_utf8
.Ve
.PP
If \f(CW$enable\fR is true (or missing), then the encode method will encode the \s-1JSON\s0 result
into \s-1UTF\-8\s0, as required by many protocols, while the decode method expects to be handled
an UTF\-8\-encoded string. Please note that UTF\-8\-encoded strings do not contain any
characters outside the range 0..255, they are thus useful for bytewise/binary I/O.
.PP
In future versions, enabling this option might enable autodetection of the \s-1UTF\-16\s0 and \s-1UTF\-32\s0
encoding families, as described in \s-1RFC4627\s0.
.PP
If \f(CW$enable\fR is false, then the encode method will return the \s-1JSON\s0 string as a (non\-encoded)
Unicode string, while decode expects thus a Unicode string. Any decoding or encoding
(e.g. to \s-1UTF\-8\s0 or \s-1UTF\-16\s0) needs to be done yourself, e.g. using the Encode module.
.PP
Example, output UTF\-16BE\-encoded \s-1JSON:\s0
.PP
.Vb 2
\&  use Encode;
\&  $jsontext = encode "UTF-16BE", JSON::XS->new->encode ($object);
.Ve
.PP
Example, decode UTF\-32LE\-encoded \s-1JSON:\s0
.PP
.Vb 2
\&  use Encode;
\&  $object = JSON::XS->new->decode (decode "UTF-32LE", $jsontext);
.Ve
.PP
See to \*(L"\s-1UNICODE\s0 \s-1HANDLING\s0 \s-1ON\s0 \s-1PERLS\s0\*(R" in \s-1JSON::PP\s0 if the backend is \s-1PP\s0.
.Sh "pretty"
.IX Subsection "pretty"
.Vb 1
\&    $json = $json->pretty([$enable])
.Ve
.PP
This enables (or disables) all of the \f(CW\*(C`indent\*(C'\fR, \f(CW\*(C`space_before\*(C'\fR and
\&\f(CW\*(C`space_after\*(C'\fR (and in the future possibly more) flags in one call to
generate the most readable (or most compact) form possible.
.PP
Equivalent to:
.PP
.Vb 1
\&   $json->indent->space_before->space_after
.Ve
.PP
The indent space length is three and \s-1JSON::XS\s0 cannot change the indent
space length.
.Sh "indent"
.IX Subsection "indent"
.Vb 1
\&    $json = $json->indent([$enable])
.Ve
.PP
.Vb 1
\&    $enabled = $json->get_indent
.Ve
.PP
If \f(CW$enable\fR is true (or missing), then the \f(CW\*(C`encode\*(C'\fR method will use a multiline
format as output, putting every array member or object/hash key-value pair
into its own line, identing them properly.
.PP
If \f(CW$enable\fR is false, no newlines or indenting will be produced, and the
resulting \s-1JSON\s0 text is guarenteed not to contain any \f(CW\*(C`newlines\*(C'\fR.
.PP
This setting has no effect when decoding \s-1JSON\s0 texts.
.PP
The indent space length is three.
With \s-1JSON::PP\s0, you can also access \f(CW\*(C`indent_length\*(C'\fR to change indent space length.
.Sh "space_before"
.IX Subsection "space_before"
.Vb 1
\&    $json = $json->space_before([$enable])
.Ve
.PP
.Vb 1
\&    $enabled = $json->get_space_before
.Ve
.PP
If \f(CW$enable\fR is true (or missing), then the \f(CW\*(C`encode\*(C'\fR method will add an extra
optional space before the \f(CW\*(C`:\*(C'\fR separating keys from values in \s-1JSON\s0 objects.
.PP
If \f(CW$enable\fR is false, then the \f(CW\*(C`encode\*(C'\fR method will not add any extra
space at those places.
.PP
This setting has no effect when decoding \s-1JSON\s0 texts.
.PP
Example, space_before enabled, space_after and indent disabled:
.PP
.Vb 1
\&   {"key" :"value"}
.Ve
.Sh "space_after"
.IX Subsection "space_after"
.Vb 1
\&    $json = $json->space_after([$enable])
.Ve
.PP
.Vb 1
\&    $enabled = $json->get_space_after
.Ve
.PP
If \f(CW$enable\fR is true (or missing), then the \f(CW\*(C`encode\*(C'\fR method will add an extra
optional space after the \f(CW\*(C`:\*(C'\fR separating keys from values in \s-1JSON\s0 objects
and extra whitespace after the \f(CW\*(C`,\*(C'\fR separating key-value pairs and array
members.
.PP
If \f(CW$enable\fR is false, then the \f(CW\*(C`encode\*(C'\fR method will not add any extra
space at those places.
.PP
This setting has no effect when decoding \s-1JSON\s0 texts.
.PP
Example, space_before and indent disabled, space_after enabled:
.PP
.Vb 1
\&   {"key": "value"}
.Ve
.Sh "relaxed"
.IX Subsection "relaxed"
.Vb 1
\&    $json = $json->relaxed([$enable])
.Ve
.PP
.Vb 1
\&    $enabled = $json->get_relaxed
.Ve
.PP
If \f(CW$enable\fR is true (or missing), then \f(CW\*(C`decode\*(C'\fR will accept some
extensions to normal \s-1JSON\s0 syntax (see below). \f(CW\*(C`encode\*(C'\fR will not be
affected in anyway. \fIBe aware that this option makes you accept invalid
\&\s-1JSON\s0 texts as if they were valid!\fR. I suggest only to use this option to
parse application-specific files written by humans (configuration files,
resource files etc.)
.PP
If \f(CW$enable\fR is false (the default), then \f(CW\*(C`decode\*(C'\fR will only accept
valid \s-1JSON\s0 texts.
.PP
Currently accepted extensions are:
.IP "* list items can have an end-comma" 4
.IX Item "list items can have an end-comma"

src/modules/casper/content/json.js
/*
Source: http://www.json.org/json.js

Copyright (c) 2005 JSON.org

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The Software shall be used for Good, not Evil.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

/*
    The global object JSON contains two methods.

    JSON.stringify(value) takes a JavaScript value and produces a JSON text.
    The value must not be cyclical.

    JSON.parse(text) takes a JSON text and produces a JavaScript value. It will
    return false if there is an error.
*/
if (typeof(Casper) == 'undefined') {
    var Casper = {};
}
Casper.JSON = function () {
    var newline = (navigator.platform == "Win32") ? "\r\n":"\n";
    var m = {
            '\b': '\\b',
            '\t': '\\t',
            '\n': '\\n',
            '\f': '\\f',
            '\r': '\\r',
            '"' : '\\"',
            '\\': '\\\\'
        },
        s = {
            'boolean': function (x, ind) {
                if (typeof ind == 'undefined') ind = 0;
                return String(x, ind);
            },
            number: function (x, ind) {
                if (typeof ind == 'undefined') ind = 0;
                return isFinite(x) ? String(x) : 'null';
            },
            string: function (x, ind) {
                if (typeof ind == 'undefined') ind = 0;
                if (/["\\\x00-\x1f]/.test(x)) {
                    x = x.replace(/([\x00-\x1f\\"])/g, function(a, b) {
                        var c = m[b];
                        if (c) {
                            return c;
                        }
                        c = b.charCodeAt();
                        return '\\u00' +
                            Math.floor(c / 16).toString(16) +
                            (c % 16).toString(16);
                    });
                }
                return '"' + x + '"';
            },
            object: function (x, ind) {
                if (typeof ind == 'undefined') ind = 0;
                if (x) {
                    var a = [], b, f, i, l, v;
                    if (x instanceof Array) {
                        a[0] = newline+indent(ind)+'[';
                        l = x.length;
                        for (i = 0; i < l; i += 1) {
                            v = x[i];
                            f = s[typeof v];
                            if (f) {
                                v = f(v, ind+1);
                                if (typeof v == 'string') {
                                    if (b) {
                                        a[a.length] = ','+newline+indent(ind);
                                    }
                                    a[a.length] = v;
                                    b = true;
                                }
                            }
                        }
                        a[a.length] = ']'+newline+indent(ind);
                    } else if (x instanceof Object) {
                        a[0] = newline+indent(ind)+'{';
                        for (i in x) {
                            v = x[i];
                            f = s[typeof v];
                            if (f) {
                                v = f(v, ind+1);
                                if (typeof v == 'string') {
                                    if (b) {
                                        a[a.length] = ','+newline+indent(ind);
                                    }
                                    a.push(s.string(i), ':', v);
                                    b = true;
                                }
                            }
                        }
                        a[a.length] = '}'+newline+indent(ind);
                    } else {
                        return null;
                    }
                    return a.join('');
                }
                return 'null';
            }
        };
    var indent = function(i) {
        var si = '  ';
        for (var x = 0; x < i; x++) { si += '  '; }
        return si;
    };
    return {
        copyright: '(c)2005 JSON.org',
        license: 'http://www.crockford.com/JSON/license.html',
/*
    Stringify a JavaScript value, producing a JSON text.
*/
        stringify: function (v) {
            var f = s[typeof v];
            if (f) {
                v = f(v);
                if (typeof v == 'string') {
                    return v;
                }
            }
            return null;
        },
/*
    Parse a JSON text, producing a JavaScript value.
    It returns false if there is a syntax error.
*/
        parse: function (text) {
            try {
                return !(/[^,:{}\[\]0-9.\-+Eaeflnr-u \n\r\t]/.test(
                        text.replace(/"(\\.|[^"\\])*"/g, ''))) &&
                    eval('(' + text + ')');
            } catch (e) {
                return false;
            }
        }
    };
}();
